Enterprise user overview page

orangejenny opened the pull request:
## Summary
https://dimagi-dev.atlassian.net/browse/USH-999

This is the first step towards a cross-domain users UI. It currently reuses most of the web users page code. It's behind a feature flag that will not be turned on for anyone external.

Probably easier to review by commit. Be aware that the first commit is almost completely copy/paste that gets re-combined later on.

## Feature Flag
USH: UI for managing all web users in an enterprise

## Product Description
![Screen Shot 2021-07-06 at 2 53 56 PM](https://user-images.githubusercontent.com/1486591/124656376-1df0f980-de6f-11eb-9633-ddc9b87d01aa.png)

## Safety Assurance

- [x] Risk label is set correctly
- [x] All migrations are backwards compatible and won't block deploy
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change
- [x] If QA is part of the safety story, the "Awaiting QA" label is used
- [x] I have confidence that this PR will not introduce a regression for the reasons below

### Automated test coverage

None

### QA Plan

Not requesting QA.

### Safety story
I'll smoke test the web users UI on staging. The new UI will be hidden from users and doesn't need to be QAed at this point.

### Rollback instructions

- [x] This PR can be reverted after deploy with no further considerations

orangejenny committed:
Added enterprise users page that just copies web users page

orangejenny committed:
Added users from throughout enterprise

orangejenny committed:
Added linked mobile users

orangejenny committed:
Added role column

orangejenny committed:
Added profile column

orangejenny committed:
Added column for other project spaces

orangejenny committed:
Added visual distinction to linked mobile users

orangejenny committed:
Removed edit links

orangejenny committed:
Added feature flag

orangejenny committed:
Extracted login_as_user_filter

orangejenny committed:
Renamed get_user_data_profile and added privilege checks

orangejenny committed:
Combined javascript back into a single model

orangejenny committed:
Extracted _get_web_users

orangejenny commented on corehq/apps/users/templates/users/enterprise_users.html:
```diff
@@ -0,0 +1,85 @@
+    <div class="panel-heading">
+      <div class="row">
+        <div class="col-sm-6">
+          <h3 class="panel-title" style="padding-top: 7px;">{% trans 'Users' %}</h3>
```
inline style copied from web_users.html

orangejenny commented on corehq/apps/users/templates/users/enterprise_users.html:
```diff
@@ -0,0 +1,85 @@
+                visible: loginAsUserCount > 0,
+              "></i>
+              <i class="fa fa-link" data-bind="visible: loginAsUser" style="margin-left: 15px;"></i>
+              <span data-bind="attr: {style: (loginAsUserCount || loginAsUser ? 'margin-left: 1px' : 'margin-left: 15px;')}, text: username"></span>
```
Didn't clean up the inline styles because the current UI (icons, spacing) isn't what's in the design anyway, so this area is likely to change.

orangejenny committed:
Fixed whitespace

orangejenny committed:
Fixed casing

snopoke commented on corehq/apps/es/users.py:
```diff
@@ -63,13 +63,18 @@ def show_only_inactive(self):
 
 
-def domain(domain, allow_mirroring=False):
+def domain(domain, allow_mirroring=False, include_mirrors=False):
```
this could use a docstring to explain what these two `mirroring` args mean

orangejenny replied:
the new param was removed later

snopoke commented on corehq/apps/es/users.py:
```diff
@@ -63,18 +64,16 @@ def show_only_inactive(self):
-            domains.append(source_domain)
+        return domains([domain, source_domain])
+    else:
+        return domains([domain])
```
this doesn't need to be a list

snopoke commented on corehq/apps/users/views/__init__.py:
```diff
@@ -741,19 +741,9 @@ def create_domain_permission_mirror(request, domain):
-        .start(skip).size(limit).run()
-    )
-    web_users = [WebUser.wrap(w) for w in web_result.hits]
+    (web_users, pagination) = _get_web_users(request, [domain])
```
`[domain]` should be `domains`

nit: tuple brackets are unnecessary

```suggestion
    web_users, pagination = _get_web_users(request, domains)
```

snopoke commented on corehq/apps/users/views/__init__.py:
```diff
@@ -723,19 +740,67 @@ def create_domain_permission_mirror(request, domain):
+@require_can_edit_or_view_web_users
+@require_GET
+def paginate_web_users(request, domain):
+    (web_users, pagination) = _get_web_users(request, [domain])
```
nit:
```suggestion
    web_users, pagination = _get_web_users(request, [domain])
```

snopoke reviewed (changes requested):
One 'bug' and the rest just small comments

orangejenny committed:
Fixed _get_web_users param and cleaned up parens

orangejenny commented:
@snopoke Thanks for review, updated in d62173d4d23

orangejenny committed:
Updated to not show any domains that aren't in the enterprise

orangejenny commented:
Tested on staging and fixed a minor bug. Merging.

orangejenny commented:
...didn't notice all the test failures. Not merging. Fixing those.

orangejenny committed:
Removed new privilege check

It isn't strictly necessary for this PR, and it breaks a bunch of tests
that would need to have domain_has_privilege patched.

orangejenny committed:
Removed unused imports

orangejenny committed:
Fixed accessor domains logic so None is never passed to ES

orangejenny committed:
Fixed previous commit

biyeun committed:
Merge branch 'master' of github.com:dimagi/commcare-hq into jls/enterprise-user-overview
